src/engine/renderer/tr_shade.cpp

```diff
@@ -38,25 +38,96 @@ static void EnableAvailableFeatures()
 	glConfig2.dynamicLight = r_dynamicLight.Get();
 	glConfig2.staticLight = r_staticLight.Get();
 
-	if ( r_dynamicLightRenderer.Get() == Util::ordinal( dynamicLightRenderer_t::TILED )
-		&& !glConfig2.textureFloatAvailable )
+	if ( glConfig2.dynamicLight || glConfig2.staticLight )
 	{
-		if ( glConfig2.dynamicLight || glConfig2.staticLight )
+		if ( r_dynamicLightRenderer.Get() == Util::ordinal( dynamicLightRenderer_t::TILED ) )
 		{
-			Log::Warn("Tiled dynamic light renderer not used because GL_ARB_texture_float is not available.");
-		}
+			if ( !glConfig2.textureFloatAvailable )
+			{
+				Log::Warn( "Tiled dynamic light renderer disabled because GL_ARB_texture_float is not available.");
+				glConfig2.dynamicLight = false;
+				glConfig2.staticLight = false;
+			}
 
-		glConfig2.dynamicLight = false;
-		glConfig2.staticLight = false;
+			// See below about ALU instructions on ATI R300 and Intel GMA 3.
+			if ( !glConfig2.glCoreProfile && glConfig2.maxAluInstructions < 128 )
+			{
+				Log::Warn( "Tiled dynamic light rendered disabled because GL_MAX_PROGRAM_ALU_INSTRUCTIONS_ARB is too small: %d", glConfig2.maxAluInstructions );
+			}
+		}
 	}
 
 	glConfig2.shadowingMode = shadowingMode_t( r_shadows.Get() );
-
 	glConfig2.shadowMapping = glConfig2.shadowingMode >= shadowingMode_t::SHADOWING_ESM16;
 
-	if ( !glConfig2.textureFloatAvailable )
+	if ( glConfig2.shadowMapping )
 	{
-		glConfig2.shadowMapping = false;
+		if ( !glConfig2.textureFloatAvailable )
+		{
+			Log::Warn( "Shadow mapping disabled because ARB_texture_float is not available" );
+			glConfig2.shadowMapping = false;
+		}
+	}
+
+	glConfig2.deluxeMapping = r_deluxeMapping->integer;
+	glConfig2.normalMapping = r_normalMapping->integer;
+	glConfig2.specularMapping = r_specularMapping->integer;
+	glConfig2.physicalMapping = r_physicalMapping->integer;
+	glConfig2.reliefMapping = r_reliefMapping->integer;
+
+	/* ATI R300 and Intel GMA 3 only have 64 tex ALU instructions, which is not enough for some shader
+	variants. For example the lightMapping shader permutation with macros USE_GRID_LIGHTING and
+	USE_GRID_DELUXE_MAPPING from the medium graphics preset requires 67 ALU.
+	For comparison, ATI R400 and R500 have 512 of them. */
+	if ( !glConfig2.glCoreProfile && glConfig2.maxAluInstructions < 128 )
+	{
+		static const std::vector<std::pair<bool*, std::string>> aluFeatures = {
```

Comment:
Can be array rather than vector

Reply:
Array is annoying because it requires a length. 😜️

Reply:
Are you thinking of std::array? I meant a C array, which doesn't require a length.

Reply:
Yeah, I was thinking of shiny C++ `std:array`. I added a commit, good old C to the rescue!